File: frontend/lib/models/global_search_response_dto.dart
import 'part_dto.dart';

class GlobalSearchResponseDto {
  final List<PartDto> parts;
  final List<SearchMaintenanceRequestDto> maintenanceRequests;
  final List<SearchWorkLogDto> workLogs;
  final List<SearchClubDto> clubs;

  GlobalSearchResponseDto({
    required this.parts,
    required this.maintenanceRequests,
    required this.workLogs,
    required this.clubs,
  });

  factory GlobalSearchResponseDto.fromJson(Map<String, dynamic> json) {
    final partsJson = json['parts'] as List<dynamic>? ?? const [];
    final requestsJson = json['maintenanceRequests'] as List<dynamic>? ?? const [];
    final workLogsJson = json['workLogs'] as List<dynamic>? ?? const [];
    final clubsJson = json['clubs'] as List<dynamic>? ?? const [];

    return GlobalSearchResponseDto(
      parts: partsJson
          .whereType<Map>()
          .map((e) => PartDto.fromJson(Map<String, dynamic>.from(e as Map)))
          .toList(),
      maintenanceRequests: requestsJson
          .whereType<Map>()
          .map((e) => SearchMaintenanceRequestDto.fromJson(Map<String, dynamic>.from(e as Map)))
          .toList(),
      workLogs: workLogsJson
          .whereType<Map>()
          .map((e) => SearchWorkLogDto.fromJson(Map<String, dynamic>.from(e as Map)))
          .toList(),
      clubs: clubsJson
          .whereType<Map>()
          .map((e) => SearchClubDto.fromJson(Map<String, dynamic>.from(e as Map)))
          .toList(),
    );
  }
}

class SearchMaintenanceRequestDto {
  final int id;
  final String? status;
  final String? clubName;
  final int? laneNumber;
  final String? mechanicName;
  final DateTime? requestedAt;

  SearchMaintenanceRequestDto({
    required this.id,
    this.status,
    this.clubName,
    this.laneNumber,
    this.mechanicName,
    this.requestedAt,
  });

  factory SearchMaintenanceRequestDto.fromJson(Map<String, dynamic> json) {
    DateTime? parseDate(dynamic value) =>
        (value is String && value.isNotEmpty) ? DateTime.tryParse(value) : null;

    return SearchMaintenanceRequestDto(
      id: (json['id'] as num).toInt(),
      status: json['status'] as String?,
      clubName: json['clubName'] as String?,
      laneNumber: (json['laneNumber'] as num?)?.toInt(),
      mechanicName: json['mechanicName'] as String?,
      requestedAt: parseDate(json['requestedAt']),
    );
  }
}

class SearchWorkLogDto {
  final int id;
  final String? status;
  final String? workType;
  final String? clubName;
  final int? laneNumber;
  final String? mechanicName;
  final String? problemDescription;
  final DateTime? createdAt;

  SearchWorkLogDto({
    required this.id,
    this.status,
    this.workType,
    this.clubName,
    this.laneNumber,
    this.mechanicName,
    this.problemDescription,
    this.createdAt,
  });

  factory SearchWorkLogDto.fromJson(Map<String, dynamic> json) {
    DateTime? parseDate(dynamic value) =>
        (value is String && value.isNotEmpty) ? DateTime.tryParse(value) : null;

    return SearchWorkLogDto(
      id: (json['id'] as num).toInt(),
      status: json['status'] as String?,
      workType: json['workType'] as String?,
      clubName: json['clubName'] as String?,
      laneNumber: (json['laneNumber'] as num?)?.toInt(),
      mechanicName: json['mechanicName'] as String?,
      problemDescription: json['problemDescription'] as String?,
      createdAt: parseDate(json['createdAt']),
    );
  }
}

class SearchClubDto {
  final int id;
  final String? name;
  final String? address;
  final bool? active;
  final bool? verified;

  SearchClubDto({
    required this.id,
    this.name,
    this.address,
    this.active,
    this.verified,
  });

  factory SearchClubDto.fromJson(Map<String, dynamic> json) {
    return SearchClubDto(
      id: (json['id'] as num).toInt(),
      name: json['name'] as String?,
      address: json['address'] as String?,
      active: json['active'] as bool?,
      verified: json['verified'] as bool?,
    );
  }
}

File: backend/src/main/java/ru/bowling/bowlingapp/DTO/GlobalSearchResponseDTO.java
package ru.bowling.bowlingapp.DTO;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.time.LocalDateTime;
import java.util.List;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class GlobalSearchResponseDTO {

    @Builder.Default
    private List<PartDto> parts = List.of();

    @Builder.Default
    private List<MaintenanceRequestResult> maintenanceRequests = List.of();

    @Builder.Default
    private List<WorkLogResult> workLogs = List.of();

    @Builder.Default
    private List<ClubResult> clubs = List.of();

    @Data
    @Builder
    @NoArgsConstructor
    @AllArgsConstructor
    public static class MaintenanceRequestResult {
        private Long id;
        private String status;
        private String clubName;
        private Integer laneNumber;
        private String mechanicName;
        private LocalDateTime requestedAt;
    }

    @Data
    @Builder
    @NoArgsConstructor
    @AllArgsConstructor
    public static class WorkLogResult {
        private Long id;
        private String status;
        private String workType;
        private String clubName;
        private Integer laneNumber;
        private String mechanicName;
        private String problemDescription;
        private LocalDateTime createdAt;
    }

    @Data
    @Builder
    @NoArgsConstructor
    @AllArgsConstructor
    public static class ClubResult {
        private Long id;
        private String name;
        private String address;
        private Boolean active;
        private Boolean verified;
    }
}
